### PR TITLE
SYS-291 : Fix log endpoint auth

### PR DESCRIPTION
**Pull request by asyed94**
Adds the same auth to log stats endpoint on json-rpc-server as debug endpoints on shardeum validators.

- adds rate limiting to debug endpoints (set to 100 calls per 15 min)

### CODE REVIEW

**Comment by chrypnotoad** [comment]
@CodiumAI-Agent /review

**Comment by chrypnotoad** [comment]
depending on merge order this is going to require changes to the (not yet merged) jest tests #43

File: src/middlewares/debugMiddleware.ts [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+          //     const parsedCounter = parseInt(sigObj.count)
+          //     if (Number.isNaN(parsedCounter)) {
+          //       console.log('Counter is not a number')
+          //     } else {
+          //       console.log('Counter is not larger than last counter', parsedCounter, lastCounter)
+          //     }
+          //   }
+        }
+      }
+    }
```
Comment (chrypnotoad) [inline src/middlewares/debugMiddleware.ts]: do we need these other else's here? Is this unfinished or was this in progress stuff that needs to be deleted?

File: src/routes/log.ts [hunk 2]
```diff
@@ -371,6 +371,12 @@ router.route('/stopRPCCapture').get(async function (req: Request, res: Response)
   CONFIG.statLog = false
   res.json({ message: 'RPC interface recording disabled' }).status(200)
 })
+router.route('/countRPCCapture').get(async function (req: Request, res: Response) {
+  if (!CONFIG.statLog) {
+    return res.json({ message: 'Interface stats recording disabled' }).status(304)
```
Comment (chrypnotoad) [inline src/routes/log.ts]: 304 status code is really intended to indicate the resource hasn't changed generally around caching. 
I think 200 OK or 204 No Content would make more sense

**Review by chrypnotoad — CHANGES_REQUESTED** [review]

File: src/routes/log.ts [hunk 3]
```diff
@@ -2,15 +2,15 @@ import { db } from '../storage/sqliteStorage'
 import express, { Request, Response } from 'express'
 export const router = express.Router()
 import { CONFIG } from '../config'
-import { debug_info } from '../logger'
+import { debug_info, getInterfaceStatCounts } from '../logger'
```
Comment (chrypnotoad) [inline src/routes/log.ts]: this getInterfaceStatCounts does not exist

File: src/routes/log.ts [hunk 4]
```diff
@@ -365,12 +365,18 @@ router.route('/startRPCCapture').get(async function (req: Request, res: Response
...
   debug_info.interfaceRecordingEndTime = Date.now()
   CONFIG.statLog = false
   res.json({ message: 'RPC interface recording disabled' }).status(200)
 })
+router.route('/countRPCCapture').get(async function (req: Request, res: Response) {
+  if (!CONFIG.statLog) {
+    return res.json({ message: 'Interface stats recording disabled' }).status(204)
+  }
+  res.json(getInterfaceStatCounts()).status(200)
+})
```
Comment (chrypnotoad) [inline src/routes/log.ts]: I don't see this method mentioned in the readme. Look's like it was added in this PR and is calling a function that doesn't exist.

**Review by chrypnotoad — CHANGES_REQUESTED** [review]

File: src/types.ts [hunk 5]
```diff
@@ -383,3 +383,11 @@ export type RequestTimeout = {
...
   full_nodelist: number
 }
+
+// Copied from shardus-core src/shardus/shardus-types.ts
+export declare enum DevSecurityLevel {
+  Unauthorized = 0,
+  Low = 1,
+  Medium = 2,
+  High = 3
+}
```
Comment (chrypnotoad) [inline src/types.ts]: I copied this from `shardus-core` should we bring it in as a dependency instead?

File: src/logger.ts [hunk 6]
```diff
@@ -145,7 +139,6 @@ export function setupLogEvents(): void {
     }
     txStatusSaver(detailedList)
   })
-  /* eslint-enable security/detect-object-injection */
```
Comment (chrypnotoad) [inline src/logger.ts]: Was this intentionally removed?

File: src/logger.ts [hunk 7]
```diff
@@ -38,72 +38,66 @@ export const debug_info = {
...
-  } catch (e) {
-    console.log(e)
   }
 
   apiPerfLogData = []
-  apiPerfLogTicket = {}
 }
 
 export function setupLogEvents(): void {
-  /* eslint-disable security/detect-object-injection */
```
Comment (chrypnotoad) [inline src/logger.ts]: `apiPerfLogTicket[ticket]` causes the object-injection warning and is still used in this function, was this intentionally deleted?

File: src/middlewares/debugMiddleware.ts [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+        if (currentCounter > lastCounter && currentCounter <= currentTime + MAX_COUNTER_BUFFER_MILLISECONDS) {
+          let verified = verify(sigObj, ownerPk)
+          if (verified === true) {
+            const authorized = ensureKeySecurity(ownerPk, authLevel)
+            if (authorized) {
+              lastCounter = currentCounter
+              next()
+              return
+            } else {
+              //   /* prettier-ignore */ if (logFlags.verbose) console.log('Authorization failed for security level', authLevel)
```
Comment (chrypnotoad) [inline src/middlewares/debugMiddleware.ts]:
```suggestion
```
delete unused log

File: src/middlewares/debugMiddleware.ts [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+                status: 403,
+                message: 'FORBIDDEN!',
+              })
+            }
+          }
+        }
+      }
+    }
+  } catch (error) {
+    // /* prettier-ignore */ if (logFlags.verbose) console.log('Error in handleDebugAuth:', error)
```
Comment (chrypnotoad) [inline src/middlewares/debugMiddleware.ts]:
```suggestion
```
delete unused log

**Review by chrypnotoad — DISMISSED** [review]

File: src/server.ts [hunk 10]
```diff
@@ -103,24 +102,20 @@ if (config.dashboard.enabled && config.dashboard.dist_path) {
...
-  const ip = req.ip || '127.0.0.1'
-  const port = req.connection.localPort || 9001
-  const success = changeNode(ip, port, true)
-  if (!success) {
-    res.end(`Ip not in the nodelist ${ip}:${port}, node subscription rejected`)
-    return
-  }
+  const ip = query.ip.toString() || '127.0.0.1'
+  const port = parseInt(query.port.toString()) || 9001
+  changeNode(ip, port)
```
Comment (mssabr01) [inline src/server.ts]: This function can fail and this should be handled. The validation of the input is done in changeNode. If ip is not a valid IP address then nothing happens.

File: src/middlewares/debugMiddleware.ts [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+    const pkClearance = devPublicKeys[pubKey]
+    return pkClearance !== undefined && pkClearance >= level
+}
+
+export function getDevPublicKeys(): {[pubkey:string]: number}  {
+    return CONFIG.devPublicKeys || {};
+}
+
+// This function is used to check if the request is authorized to access the debug endpoint
+export function handleDebugAuth(_req: any, res: any, next: any, authLevel: any)
```
Comment (mssabr01) [inline src/middlewares/debugMiddleware.ts]: This middleware or the endpoints that call it need a rate limiter. See service points in the validator repo for example.

**Review by mssabr01 — CHANGES_REQUESTED** [review]

**Review by mssabr01 — DISMISSED** [review]

File: src/middlewares/debugMiddleware.ts [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+    if (_req.query.sig != null && _req.query.sig_counter != null) {
+      const devPublicKeys = getDevPublicKeys() // This should return list of public keys
+      const requestSig = _req.query.sig
+      // Check if signature is valid for any of the public keys
+      for (const ownerPk in devPublicKeys) {
+        let sigObj = {
+          route: _req.route.path,
+          count: String(_req.query.sig_counter),
+          sign: { owner: ownerPk, sig: requestSig },
+        }
```
Comment (kun6fup4nd4) [inline src/middlewares/debugMiddleware.ts]: signing path w/o params is risky. in this exact implementation no vulnerability cuz of counter

**Review by kun6fup4nd4 — DISMISSED** [review]

**Review by chrypnotoad — DISMISSED** [review]

**Comment by chrypnotoad** [comment]
@CodiumAI-Agent /review

File: src/server.ts [hunk 13]
```diff
@@ -35,6 +33,9 @@ import { setupArchiverDiscovery } from '@shardus/archiver-discovery'
 import { setDefaultResultOrder } from 'dns'
 import { nestedCountersInstance } from './utils/nestedCounters'
 import { methodWhitelist } from './middlewares/methodWhitelist'
+import { isDebugModeMiddlewareLow, rateLimitedDebugAuth } from './middlewares/debugMiddleware'
+import { isIP, isIPv4 } from 'net';
```
Comment (chrypnotoad) [inline src/server.ts]:
```suggestion
import { isIPv4 } from 'net'; 
```
remove unused import

**Review by chrypnotoad — DISMISSED** [review]

**Review by timadinorth — APPROVED** [review]

**Review by kun6fup4nd4 — DISMISSED** [review]

**Pull request closed without merging**